app/src/lib/stores/commit-status-store.ts

```diff
@@ -261,13 +292,46 @@ export class CommitStatusStore {
       // notifying subscribers we ensure they keep their current status
       // if they have one and that we attempt to fetch it again on the same
       // schedule as the others.
-      log.debug(`Failed fetching status for ref ${ref} (${owner}/${name})`, err)
-
       const existingEntry = this.cache.get(key)
-      const status = existingEntry === undefined ? null : existingEntry.status
+      const check = existingEntry?.check ?? null
+
+      this.cache.set(key, { check, fetchedAt: new Date() })
+      return
+    }
+
+    if (statuses !== null) {
+      checks.push(...statuses.statuses.map(apiStatusToRefCheck))
+    }
+
+    if (checkRuns !== null) {
+      checks.push(...checkRuns.check_runs.map(apiCheckRunToRefCheck))
+    }
 
-      this.cache.set(key, { status, fetchedAt: new Date() })
+    if (checks.length === 0) {
+      // This case is distinct from when we fail to call the API in
+      // that this means there are no checks or statuses so we should
+      // clear whatever info we've got for this ref.
+      this.cache.set(key, { check: null, fetchedAt: new Date() })
+      return
+    }
+
+    let status: APICheckStatus
+    let conclusion: APICheckConclusion | null = null
+
+    if (checks.some(isIncompleteOrFailure)) {
+      status = 'completed'
+      conclusion = 'failure'
```

Comment:
Super-small-nitpick: can we move the logic in `isIncomplete()` into the `isFailure()` method and have only that method?

While I understand the distinction between the two, I think here it creates more confusion than clarity, since `isIncomplete` can easily be confused with the `complete` status on the API (when reading this I initially thought that it made no sense to set the combined status to "complete" when there's any incomplete step 😆).

Reply:
I absolutely agree but in this case I've chosen to mirror the logic on the dotcom side so I'd prefer to keep the terminology and logic in sync in order to more easily be able to adapt should they decide to support more states in the future.

Reply:
> I absolutely agree but in this case I've chosen to mirror the logic on the dotcom side so I'd prefer to keep the terminology and logic in sync in order to more easily be able to adapt should they decide to support more states in the future.

I'm not sure I follow, you mean there's also an `isIncomplete()` method on dotcom that only returns `true` when status=='complete'`? If that's the case then I agree about keeping it that way (note that my suggestion was only stylistic and didn't imply changing the logic).

Reply:
> I'm not sure I follow, you mean there's also an `isIncomplete()` method on dotcom that only returns `true` when status=='complete'`?

Right, an incomplete check run is a check run which has status 'complete' but doesn't have a definite conclusion (success, failure, neutral) because it's been cancelled, timed out, or marked as stale. So an incomplete check is different from an in progress check or a queued check. I've added a comment to hopefully clarify this in the code.